refactor: update files to be compatible with Scala 3

jaschdoc opened the pull request:
Related to https://github.com/flix/flix/pull/8286

magnus-madsen commented:
@mlutze @JonathanStarup Do you see any changes you do not "agree with"? I.e. feature loss that is unacceptable?

magnus-madsen commented:
@jaschdoc The use of implicit apply is gone it seems? @JonathanStarup was arguing to have that in Flix...

JonathanStarup commented:
hmm i think thats a simplified summary 🤔

mlutze commented on main/src/ca/uwaterloo/flix/language/dbg/DocAstFormatter.scala:
```diff
@@ -22,6 +22,7 @@ import ca.uwaterloo.flix.language.dbg.DocAst.Expr._
 import ca.uwaterloo.flix.language.dbg.DocAst._
 
 import scala.annotation.tailrec
+import scala.collection.immutable.::
```
Seems redundant

jaschdoc replied:
See previous comment

mlutze commented on main/src/ca/uwaterloo/flix/language/dbg/DocAstFormatter.scala:
```diff
@@ -324,9 +325,9 @@ object DocAstFormatter {
         val (curriedArgs, res) = collectArrowTypes(arrow)
         val formattedArgs = curriedArgs.map {
-          case ts@(Type.Tuple(_) :: Nil) =>
+          case ts@Type.Tuple(_) :: Nil =>
```
So `::` has higher precedence than `@`? This confuses me.

jaschdoc replied:
Scala complained that the parentheses were redundant

magnus-madsen replied:
Another reason not to use @

mlutze replied:
What was the first reason?

magnus-madsen replied:
I dont like it :)

JonathanStarup commented:
but it looks fine, its mostly the editor behavior thats worrying i think. but maybe its fine since we dont use fancy features

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/dbg/DocAstFormatter.scala:
```diff
@@ -22,6 +22,7 @@ import ca.uwaterloo.flix.language.dbg.DocAst.Expr._
 import ca.uwaterloo.flix.language.dbg.DocAst._
 
 import scala.annotation.tailrec
+import scala.collection.immutable.::
```
Wait, this should not be necessary?

jaschdoc replied:
Scala couldn't figure out that this belonged to List and not DocAst

magnus-madsen replied:
Can we rename our Cons to FCons?

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Reducer.scala:
```diff
@@ -200,7 +200,7 @@ object Reducer {
     */
   private def typesOfEffect(e: Effect): Set[MonoType] = e.ops.toSet.map {
-    op: Op =>
+    (op: Op) =>
```
Does it work without parens, but with braces?

jaschdoc replied:
Can try

jaschdoc replied:
It's already inside a call to `map` with braces.

jaschdoc replied:
I can drop the type annotation, though.

magnus-madsen replied:
![Uploading CleanShot 2024-08-11 at 11.30.00@2x.png…]()

jaschdoc replied:
Nevermind, the type annotation was required, but I moved it to separate function.

mlutze reviewed (commented):
I'm not concerned about these changes specifically, but I am concerned about Scala 3. In my experience, IntelliJ support is weaker and I am strongly opposed to indentation-sensitive syntax.

magnus-madsen commented:
> I'm not concerned about these changes specifically, but I am concerned about Scala 3. In my experience, IntelliJ support is weaker and I am strongly opposed to indentation-sensitive syntax.

Agreed. We will not switch without a major discussion. But if we can be compatible then we can at least check it out.

I would also wait some years before switching when tooling has caught up / the language has stabilized (?)

jaschdoc commented:
Btw all of these changes were required by Scala and not IntelliJ. I ran the gradle from the terminal

jaschdoc commented:
I think the explicit apply method goes away when we move to enums instead of case classes

magnus-madsen commented:
@jaschdoc There are two things to fix/try to fix and then we can merge.

mlutze commented:
> @jaschdoc There are two things to fix/try to fix and then we can merge.

Sorry :P

magnus-madsen commented:
Is there a flag for scala 2.13 that enables forward compatability?

jaschdoc commented:
> Is there a flag for scala 2.13 that enables forward compatability?

I can look into it but they have a version of Scala 3 (which I think I'm using) intended for Scala 2 migration

magnus-madsen commented:
The Scala 2.13 compiler issues helpful migration warnings with the -Xsource:3 flag.

Before moving to the Scala 3 compiler, it’s recommended to enable this flag in Scala 2 and address the new warnings.

https://docs.scala-lang.org/scala3/guides/migration/tooling-scala2-xsource3.html

jaschdoc commented:
There are more changes incoming (the Scala 3 compiler was not quite satisfied). I will add those to this PR and then resolve the errors from the Scala 2 compiler in migration mode.

magnus-madsen commented:
We definitely don't want significant whitespace. The goal here is to find the minimum number of changes.

magnus-madsen commented:
Hopefully this PR can be kept less than 500 lines. Otherwise we need to rethink.

magnus-madsen commented:
@JonathanStarup @mlutze let us all review it to: (1) learn about the changes and (2) see if we are unhappy with anything.

If you have no comment, you can just post a thumbs-up.

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/CodeActionContext.scala:
```diff
@@ -17,7 +17,8 @@ package ca.uwaterloo.flix.api.lsp
 import ca.uwaterloo.flix.util.Result.{Err, Ok}
-import org.json4s.JsonAST._
+import org.json4s.jvalue2monadic
+import org.json4s.JsonAST.*
```
Do we have to use *? I think we should stick with _ to remain Scala 2.0

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:
```diff
@@ -2350,7 +2351,7 @@ object Resolver {
       case caze :: _ => Result.Ok(caze)
       // Case 2: Multiple matches. Error
-      case cazes => throw InternalCompilerException(s"unexpected duplicate tag: '$qname'.", qname.loc)
+      case null => throw InternalCompilerException(s"unexpected duplicate tag: '$qname'.", qname.loc)
```
Something is wrong here!

magnus-madsen replied:
The original case should have been: caze :: Nil not caze :: _

magnus-madsen replied:
(Sidenote: We never use null. )

jaschdoc replied:
Hmm... true, but the Scala 3 compiler said that `cazes` was unreachable unless the match exp were `null`

magnus-madsen replied:
Here the original code is broken and should be fixed, i.e. the match before the wildcard should be caze :: Nil

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:
```diff
@@ -2368,7 +2369,7 @@ object Resolver {
       case st :: _ => Result.Ok(st)
       // Case 2: Multiple matches. Error
-      case sts => throw InternalCompilerException(s"unexpected duplicate struct: '$qname'.", qname.loc)
+      case null => throw InternalCompilerException(s"unexpected duplicate struct: '$qname'.", qname.loc)
```
same comment

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/unification/CaseSetUnification.scala:
```diff
@@ -235,12 +235,12 @@ object CaseSetUnification {
     case Or(tpe1, tpe2) => Nnf.Union(nnf(tpe1), nnf(tpe2))
     case And(tpe1, tpe2) => Nnf.Intersection(nnf(tpe1), nnf(tpe2))
-    case _ => throw InternalCompilerException(s"unexpected type: $t", SourceLocation.Unknown)
+    case null => throw InternalCompilerException(s"unexpected type: $t", SourceLocation.Unknown)
```
This does not make sense to me. Is the pattern already exhaustive?

jaschdoc replied:
Unreachable again as above

magnus-madsen replied:
In this case then remove the throw

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/unification/CaseSetUnification.scala:
```diff
@@ -260,7 +260,7 @@ object CaseSetUnification {
       nnf(mkNot(tpe2))
     )
-    case _ => throw InternalCompilerException(s"unexpected type: $t", SourceLocation.Unknown)
+    case null => throw InternalCompilerException(s"unexpected type: $t", SourceLocation.Unknown)
```
Same comment.

magnus-madsen commented on main/src/ca/uwaterloo/flix/tools/SocketServer.scala:
```diff
@@ -23,10 +23,11 @@ import ca.uwaterloo.flix.util._
 import org.json4s.ParserUtil.ParseException
 import org.json4s.native.JsonMethods
-import org.json4s.native.JsonMethods._
+import org.json4s.native.JsonMethods.*
```
Revert the two uses of *.

magnus-madsen commented on main/src/ca/uwaterloo/flix/tools/Tester.scala:
```diff
@@ -148,7 +148,7 @@ object Tester {
             finished = true
 
-          case _ => // nop
+          case null => // nop
```
I am confused again.

magnus-madsen commented on main/test/ca/uwaterloo/flix/util/FlixSuite.scala:
```diff
@@ -21,7 +21,7 @@ import ca.uwaterloo.flix.runtime.{CompilationResult, TestFn}
 
 import java.nio.file.{Files, Path, Paths}
-import scala.jdk.CollectionConverters._
+import scala.jdk.CollectionConverters.*
```
Undo

magnus-madsen reviewed (commented):
I added some comments. Two things standout:

- Let us not use * for imports
- Strange things are going on with null

jaschdoc commented:
> We definitely don't want significant whitespace. The goal here is to find the minimum number of changes.

Funny thing is, if you disable significant whitespace it complains that things aren't formatted properly :thinking:

magnus-madsen commented:
> > We definitely don't want significant whitespace. The goal here is to find the minimum number of changes.
> 
> Funny thing is, if you disable significant whitespace it complains that things aren't formatted properly 🤔

WHAT?

jaschdoc commented:
> > > We definitely don't want significant whitespace. The goal here is to find the minimum number of changes.
> 
> > 
> 
> > Funny thing is, if you disable significant whitespace it complains that things aren't formatted properly 🤔
> 
> 
> 
> WHAT?

Haha yes then white space doesn't matter but it's still opinionated as to how it should be formatted

magnus-madsen commented:
As in it fails compilation?

magnus-madsen commented:
Should we try to get this merged in? We can do Resolver in a subsequent PR if that helps.

jaschdoc commented:
> As in it fails compilation?

I believe it's a warning, but we turn warnings into errors

magnus-madsen commented:
RFM?

magnus-madsen commented:
Wait. We need to deal with the case null =>

I don't want them. Can the cases be deleted or reverted back to _?

jaschdoc commented:
Still not ready. Just fixing up the last tidbits. I'll let you know :)

jaschdoc commented on main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:
```diff
@@ -2362,9 +2363,9 @@ object Resolver {
       case Nil => Result.Err(ResolutionError.UndefinedStruct(qname, qname.loc))
       // Case 1: Exactly one match. Success.
-      case st :: _ => Result.Ok(st)
+      case st :: Nil => Result.Ok(st)
```
Either we have discovered a bug in the compiler or the comments do not match what is actually intended here.

jaschdoc commented:
@magnus-madsen I think I need your help on the comment above

jaschdoc commented on main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:
```diff
@@ -2977,7 +2978,7 @@ object Resolver {
         val ns = prefix ::: qname0.namespace.parts.tail
         val qname = Name.mkQName(ns, qname0.ident.name, SourceLocation.Unknown)
-        root.symbols.getOrElse(qname.namespace, Map.empty).get(qname.ident.name)
+        root.symbols.getOrElse(qname.namespace, Map.empty).get(qname.ident.name).map(_.distinct)
```
This change may be significant?

magnus-madsen replied:
Why the extra map?

jaschdoc replied:
Same as the other comment but for consistency. I did not observe any change in behavior on this line though so this could be removed maybe?

jaschdoc commented on main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:
```diff
@@ -2949,20 +2950,20 @@ object Resolver {
+      val rootNames = root.symbols.getOrElse(Name.RootNS, Map.empty).getOrElse(qname.ident.name, Nil).map(Resolution.Declaration.apply)
 
-      envNames ::: localNames ::: currentNamespace ::: rootNames
+      (envNames ::: localNames ::: currentNamespace ::: rootNames).distinct
```
This change may be significant?

magnus-madsen replied:
Same question: Why distinct?

jaschdoc replied:
I inspected the results and the code above finds duplicate resolutions, e.g., it will find `Option.None` multiple times. However, the caller(s) of this function assumes that only one lookup should be returned if successful. This is (obviously) either intentional or a bug in this function but it seems like once we have found a resolution we shouldn't return it multiple times. Maybe it should be a set instead?

magnus-madsen commented:
We should ask @mlutze , but my first reaction is to leave the code unchanged: If there is a bug, we can find and fix it later.

Ideally, this PR should be a pure refactor.

jaschdoc commented:
> We should ask @mlutze , but my first reaction is to leave the code unchanged: If there is a bug, we can find and fix it later.
> 
> Ideally, this PR should be a pure refactor.

Ok sounds sensible :+1:

mlutze commented:
I agree that this should just be a pure refactor.

The distinct thing should not make a difference. The functions calling it should only be getting the first one they find, so it doesn't matter if there are duplicates. Any meaningful duplicates should have already been checked by the Namer.

jaschdoc commented:
Ah okay then the comments at the call sites are wrong

jaschdoc commented:
Alright let's try this again :smile: 
Matt convinced me, so I have removed the changes to Resolver and just kept it to a simple refactor :)

jaschdoc commented:
> Can we do those changes to lookup in another PR? Even if it _should_ be the same, it's better to be able to scrutinize those separately.

What do you mean? I only added the comment for clarity. The rest is specific to Scala 3 compatibility.

mlutze commented:
In commit aab42f6a0c89d0cf7eafdf54a41bf8f81b8d938d there appear to be semantic changes, no?

jaschdoc commented:
> In commit [aab42f6](https://github.com/flix/flix/commit/aab42f6a0c89d0cf7eafdf54a41bf8f81b8d938d) there appear to be semantic changes, no?

Added comments to the changes so you can see (via the link) what I've changed
